File: examples/java/src/main/java/com/xebia/functional/xef/java/auto/jdk21/Markets.java
package com.xebia.functional.xef.java.auto.jdk21;

import com.xebia.functional.xef.java.auto.AIScope;
import com.xebia.functional.xef.java.auto.ExecutionContext;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;

public class Markets {

    public record Market(String news, List<String> raisingStockSymbols, List<String> decreasingStockSymbols){}

    private static CompletableFuture<Void> stockMarketSummary(AIScope scope) {
        var news = "|" +
              "|Write a short summary of the stock market results given the provided context.";

        return scope.prompt(news, Market.class)
              .thenAccept(markets -> System.out.println(markets));
    }

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        try (AIScope scope = new AIScope(new ExecutionContext(Executors.newVirtualThreadPerTaskExecutor()))) {
            var dtf = DateTimeFormatter.ofPattern("dd/M/yyyy");
            var now = LocalDateTime.now();
            var currentDate = dtf.format(now);

            scope.contextScope(scope.search(currentDate + "Stock market results, raising stocks, decreasing stocks"),
                  Markets::stockMarketSummary).get();
        }
    }
}

File: example/src/main/kotlin/com/xebia/functional/langchain4k/auto/Markets.kt
package com.xebia.functional.langchain4k.auto

import com.xebia.functional.auto.ai
import com.xebia.functional.auto.getOrElse
import com.xebia.functional.tool.search
import kotlinx.serialization.Serializable
import java.text.SimpleDateFormat
import java.util.*

@Serializable
data class MarketNews(
  val news: String,
  val raisingStockSymbols: List<String>,
  val decreasingStockSymbols: List<String>
)

suspend fun main() {
  ai {
    val sdf = SimpleDateFormat("dd/M/yyyy")
    val currentDate = sdf.format(Date())
    agent(search("$currentDate Stock market results, raising stocks, decreasing stocks")) {
      val news: MarketNews = prompt(
        """|
                     |Write a short summary of the stock market results given the provided context.
                     """.trimMargin()
      )
      println(news)
    }
  }.getOrElse { println(it) }
}

